eth-reference-tests/src/referenceTest/java/tech/pegasys/teku/reference/phase0/fork_choice/ForkChoiceTestExecutor.java

```diff
@@ -0,0 +1,302 @@
+        try {
+          if (path.endsWith(".ssz")) {
+            return SimpleOffsetSerializer.deserialize(
+                Bytes.wrap(Files.readAllBytes(partPath)), clazz);
+          } else {
+            return mapper.readValue(partPath.toFile(), clazz);
+          }
+        } catch (IOException e) {
+          throw new IllegalArgumentException("Couldn't resolve " + path + ": " + e.getMessage());
+        }
+      }
+    }
+    return clazz.cast(value);
+  }
+
+  @Override
+  public void runTest(TestDefinition testDefinition) throws Throwable {
+    Path testDirectory = testDefinition.getTestDirectory();
+    Arguments arguments = parseForkChoiceFile(testDirectory).get();
+    Object[] objects = arguments.get();
+    BeaconState genesis = (BeaconState) objects[0];
+    @SuppressWarnings("unchecked")
+    List<Object> steps = (List<Object>) objects[1];
+    String testName = (String) objects[2];
+    Boolean protoArrayFC = (Boolean) objects[3];
+    runForkChoiceTests(genesis, steps, testName, protoArrayFC);
+  }
+
+  @ParameterizedTest(name = "{index}.{2} fork choice test")
+  @MethodSource("loadForkChoiceTests")
+  void runForkChoiceTests(
+      BeaconState genesis, List<Object> steps, String testName, boolean protoArrayFC) {
+    StateTransition st = new StateTransition();
+
+    EventBus eventBus = new EventBus();
+    RecentChainData storageClient = MemoryOnlyRecentChainData.create(eventBus);
+    storageClient.initializeFromGenesis(genesis);
+
+    ForkChoiceStrategy forkChoiceStrategy =
+        protoArrayFC
+            ? ProtoArrayForkChoiceStrategy.create(storageClient.getStore())
+            : new OrigForkChoiceStrategy(storageClient.getStore());
+
+    @SuppressWarnings("ModifiedButNotUsed")
+    List<SignedBeaconBlock> blockBuffer = new ArrayList<>();
+    @SuppressWarnings("ModifiedButNotUsed")
+    List<Attestation> attestationBuffer = new ArrayList<>();
+
+    for (Object step : steps) {
+      blockBuffer.removeIf(block -> processBlock(st, storageClient, block, forkChoiceStrategy));
+      attestationBuffer.removeIf(
+          attestation -> processAttestation(st, storageClient, attestation, forkChoiceStrategy));
+      if (step instanceof UnsignedLong) {
+        UpdatableStore.StoreTransaction transaction = storageClient.startStoreTransaction();
+        while (ForkChoiceUtil.get_current_slot(transaction).compareTo((UnsignedLong) step) < 0) {
+          ForkChoiceUtil.on_tick(transaction, transaction.getTime().plus(UnsignedLong.ONE));
+        }
+        assertEquals(step, ForkChoiceUtil.get_current_slot(transaction));
+        transaction.commit().join();
+      } else if (step instanceof SignedBeaconBlock) {
+        for (Attestation attestation :
+            ((SignedBeaconBlock) step).getMessage().getBody().getAttestations()) {
+          attestationBuffer.add(attestation);
+        }
+        if (!processBlock(st, storageClient, (SignedBeaconBlock) step, forkChoiceStrategy)) {
```

Comment:
What I understand is that: we try to process the block, and if it's unsuccessful, we add it to the blockBuffer, and then at the next step we try to process it again. Should we not be failing if block or attestation processing fails? I'm assuming we're not failing on purpose.

Reply:
It can be the case, that a block arrived earlier or its parent block has not arrived yet, etc.
Basically, any block or attestation can arrive and that can be a part of a test.
So, block/attestation processing error can be expected.

Reply:
I see. That makes sense 👍. Should we throw an error if there are still blocks/attestations left in the buffers after the last step? I assume we should have processed all the given objects at some point.

Reply:
It's tricky.
Normally, a non-empty buffer indicates something was wrong, but if a block or an attestation is invalid (e.g. wrong signature), it will stay in the buffer forever.
Some logic can be implemented to tell invalid msgs from early ones, however, it's related to how the spec is structured at the moment.
As there is some demand from 'formal method' guys for a more declarative spec, that can happen. At the moment, I prefer to ignore this - normally, if a message is rejected, than it typically should affect fork choice and head should be different.

I probably could improve tests and insert checks whether a message should be processed successfully, rejected as invalid or delayed. But it will require the test format change.

Reply:
Do you have invalid blocks and attestations in the test files on purpose? If not, all the attestations and blocks in the buffers (even the too early ones) would need to be consumed at some point, and we could throw an error if they are not consumed by the end of the test.

Reply:
I have invalid ones on purpose

Reply:
Hmm, a good next step could be to include the number of invalid blocks and attestations to test, and check that the number of invalid ones we perceived matches after processing, but I think that could be done later on, and it's for the best to merge this in soon.